File: test-suite-kotlin/src/main/kotlin/helloworld/GreeterService.kt
package helloworld

// tag::imports[]
import io.grpc.stub.StreamObserver
import io.micronaut.grpc.annotation.GrpcRestJsonExposed
import jakarta.inject.Singleton
import org.example.grpc.GreeterGrpc
import org.example.grpc.HelloRequest
import org.example.grpc.HelloResponse
import org.slf4j.LoggerFactory
// end::imports[]

// tag::clazz[]
@Singleton
class GreeterService : GreeterGrpc.GreeterImplBase() {

    @GrpcRestJsonExposed
    override fun sayHello(request: HelloRequest, responseObserver: StreamObserver<HelloResponse>) {
        LOG.debug("Received request: {}", request)
        val name = request.name
        val greeting = "Hello, $name"

        val reply = HelloResponse.newBuilder()
            .setGreeting(greeting)
            .build()

        responseObserver.onNext(reply)
        responseObserver.onCompleted()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(GreeterService::class.java)
    }
}
// end::clazz[]

File: test-suite-python/src/main/python/helloworld/GreeterService.py
# tag::imports[]
import logging

from io.grpc.stub import StreamObserver
from jakarta.inject import Singleton
from micronaut.grpc.annotation import GrpcRestJsonExposed
from org.example.grpc import GreeterGrpc, HelloRequest, HelloResponse

LOG = logging.getLogger(__name__)
# end::imports[]


# tag::clazz[]
@Singleton
class GreeterService(GreeterGrpc.GreeterImplBase):

    @GrpcRestJsonExposed
    def sayHello(self, request: HelloRequest, response_observer: StreamObserver[HelloResponse]) -> None:
        LOG.debug("Received request: %s", request)
        name = request.getName()
        greeting = "Hello, " + name

        reply = HelloResponse.newBuilder().setGreeting(greeting).build()

        response_observer.onNext(reply)
        response_observer.onCompleted()
# end::clazz[]
